patches/server/1056-Adding-PredicateChoice-to-Paper-API.patch

```diff
@@ -0,0 +1,212 @@
+ import it.unimi.dsi.fastutil.ints.IntList;
+ import it.unimi.dsi.fastutil.objects.Object2IntMap;
+ import it.unimi.dsi.fastutil.objects.Object2IntOpenCustomHashMap;
++import java.util.ArrayList;
+ import java.util.IdentityHashMap;
++import java.util.List;
+ import java.util.Map;
+ import java.util.concurrent.atomic.AtomicInteger;
+ import net.minecraft.core.registries.BuiltInRegistries;
+@@ -18,12 +20,12 @@ import net.minecraft.world.item.crafting.Ingredient;
+ import net.minecraft.world.item.crafting.Recipe;
+ 
+ public final class StackedContentsExtraMap {
+-
+     private final AtomicInteger idCounter = new AtomicInteger(BuiltInRegistries.ITEM.size()); // start at max vanilla stacked contents idx
+     private final Object2IntMap<ItemStack> exactChoiceIds = new Object2IntOpenCustomHashMap<>(ItemStackLinkedSet.TYPE_AND_TAG);
+     private final Int2ObjectMap<ItemStack> idToExactChoice = new Int2ObjectOpenHashMap<>();
+     private final StackedContents contents;
+     public final Map<Ingredient, IntList> extraStackingIds = new IdentityHashMap<>();
++    public final List<Ingredient> predicateChoices = new ArrayList<>();  // Adding PredicateChoice
+ 
+     public StackedContentsExtraMap(final StackedContents contents, final Recipe<?> recipe) {
+         this.exactChoiceIds.defaultReturnValue(-1);
+@@ -32,7 +34,7 @@ public final class StackedContentsExtraMap {
+     }
+ 
+     private void initialize(final Recipe<?> recipe) {
+-        if (recipe.hasExactIngredients()) {
++        if (recipe.hasSpecialIngredients()) {
+             for (final Ingredient ingredient : recipe.getIngredients()) {
+                 if (!ingredient.isEmpty() && ingredient.exact) {
+                     final net.minecraft.world.item.ItemStack[] items = ingredient.getItems();
+@@ -47,6 +49,11 @@ public final class StackedContentsExtraMap {
+                     idList.sort(IntComparators.NATURAL_COMPARATOR);
+                     this.extraStackingIds.put(ingredient, idList);
+                 }
++                // Adding PredicateChoice
++                else if(!ingredient.isEmpty() && ingredient.getItemPredicate() != null){
++                    this.predicateChoices.add(ingredient);
++                    this.extraStackingIds.put(ingredient, new IntArrayList()); // fill id list when accounting stacks
++                }
+             }
+         }
+     }
+@@ -67,6 +74,16 @@ public final class StackedContentsExtraMap {
+     }
+ 
+     public boolean accountStack(final ItemStack stack, final int count) {
++        // We are adding items that pass the predicate test.
++        for(final Ingredient predicateChoice : this.predicateChoices){
++            if(predicateChoice.test(stack)){
++                boolean isStackTypeRegistered = this.exactChoiceIds.getInt(stack) > -1;
++                final int id = this.registerExact(stack);
++                // We only want to add the stacking id to the list one time
++                if(!isStackTypeRegistered)
```

Comment:
You should just be able to check the `id` returned by `registerExact` if it is -1, it's already been added to the exactChoice map.